src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs

```diff
@@ -2034,6 +2034,694 @@ static bool Set(out int x)
+        }
+
+        [Fact]
+        public void NullCoalescing_ConditionalLeft()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x, y;
+        _ = (bool?)(b && c1.M(x = y = 0)) ?? false
+            ? x.ToString() // 1
+            : y.ToString(); // 2
+    }
+
+    void M2(C c1, bool b)
+    {
+        int x, y;
+        _ = (bool?)c1.M(x = y = 0) ?? false
+            ? x.ToString()
+            : y.ToString();
+    }
+
+    void M3(C c1, bool b)
+    {
+        int x, y;
+        _ = (bool?)((y = 0) is 0 && c1.M(x = 0)) ?? false
+            ? x.ToString() // 3
+            : y.ToString();
+    }
+
+    bool M(object obj) { return true; }
+}
+";
+            // Note that in definite assignment we unsplit any conditional state after visiting the left side of `??`.
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(8, 15),
+                // (9,15): error CS0165: Use of unassigned local variable 'y'
+                //             : y.ToString(); // 2
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(9, 15),
+                // (24,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 3
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(24, 15));
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_Throw()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        _ = c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    bool M(object obj) { return true; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_Cast()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        _ = (object)c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_ImplicitReferenceConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? (object)(x = 0);
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_BoxingConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? c1.M(x = 0);
```

Comment:
I'm not really sure what this is testing: the boxing conversion should occur on the entire coalescing, not on either arm.

Reply:
> the boxing conversion should occur on the entire coalescing, not on either arm.

Sorry, I didn't understand this suggestion. Based on the bound tree dump it seems like the boxing conversion happens on the entire coalescing:

```
│ │ ├─initializerOpt
│ │ │ └─conversion
│ │ │   ├─operand
│ │ │   │ └─nullCoalescingOperator
│ │ │   │   ├─leftOperand
│ │ │   │   │ └─conditionalAccess
│ │ │   │   │   ├─...
│ │ │   │   ├─rightOperand
│ │ │   │   │ └─call
│ │ │   │   │   ├─...
│ │ │   │   ├─leftConversion: Identity
│ │ │   │   ├─operatorResultKind: LeftType
│ │ │   │   ├─type: bool?
│ │ │   │   ├─isSuppressed: False
│ │ │   │   └─hasErrors: False
│ │ │   ├─conversion: Boxing
│ │ │   ├─isBaseConversion: False
│ │ │   ├─@checked: False
│ │ │   ├─explicitCastInCode: False
│ │ │   ├─constantValueOpt
│ │ │   ├─conversionGroupOpt
│ │ │   ├─originalUserDefinedConversionsOpt: {}
│ │ │   ├─type: object
│ │ │   ├─isSuppressed: False
│ │ │   └─hasErrors: False
│ │ ├─argumentsOpt
│ │ ├─inferredType: False
│ │ └─hasErrors: False
```

Now that I actually understand how implicit conversions on the left are applied, though, it feels like it's not necessary to have this test.

Reply:
> Now that I actually understand how implicit conversions on the left are applied, though, it feels like it's not necessary to have this test.

Yeah, that's the point I was trying to make. I didn't understand what it was testing.